test/quic_stream_test.c

```diff
@@ -321,9 +321,163 @@ static int test_bulk(int idx)
+                                             &readbytes, &fin))
+        || !TEST_false(fin)
+        || !TEST_size_t_eq(readbytes, 15)
+        || !TEST_mem_eq(buf, 15, simple_data, 15)
+        || !TEST_true(ossl_quic_rstream_queue_data(rstream, NULL,
+                                                   15,
+                                                   simple_data + 15,
+                                                   sizeof(simple_data) - 15, 1))
+        || !TEST_true(ossl_quic_rstream_available(rstream, &avail, &fin))
+        || !TEST_true(fin)
+        || !TEST_size_t_eq(avail, sizeof(simple_data))
+        || !TEST_true(ossl_quic_rstream_read(rstream, buf, 12,
+                                             &readbytes, &fin))
+        || !TEST_false(fin)
+        || !TEST_size_t_eq(readbytes, 12)
+        || !TEST_mem_eq(buf, 12, simple_data, 12)
+        || !TEST_true(ossl_quic_rstream_read(rstream, buf + 12, sizeof(buf) - 12,
+                                             &readbytes, &fin))
+        || !TEST_true(fin)
+        || !TEST_size_t_eq(readbytes, sizeof(buf) - 12)
+        || !TEST_mem_eq(buf, sizeof(buf), simple_data, sizeof(simple_data))
+        || !TEST_true(ossl_quic_rstream_read(rstream, buf, sizeof(buf),
+                                             &readbytes, &fin))
+        || !TEST_true(fin)
+        || !TEST_size_t_eq(readbytes, 0))
+        goto err;
+
+    ret = 1;
+
+ err:
+    ossl_quic_rstream_free(rstream);
+    return ret;
+}
+
+static int test_rstream_random(int idx)
+{
+    unsigned char *bulk_data = NULL;
+    unsigned char *read_buf = NULL;
+    QUIC_RSTREAM *rstream = NULL;
+    size_t i, read_off, queued_min;
+    const size_t data_size = 10000;
+    int r, s, fin;
+    int ret = 0;
+    size_t readbytes = 0;
+
+    if (!TEST_ptr(bulk_data = OPENSSL_malloc(data_size))
+        || !TEST_ptr(read_buf = OPENSSL_malloc(data_size))
+        || !TEST_ptr(rstream = ossl_quic_rstream_new(NULL, NULL, NULL)))
+        goto err;
+
+    for (i = 0; i < data_size; ++i)
+        bulk_data[i] = (unsigned char)(test_random() & 0xFF);
+
+    read_off = queued_min = 0;
+    for (r = 0; r < 100; ++r) {
+        for (s = 0; s < 10; ++s) {
+            size_t off = (r * 10 + s) * 10, size = 10;
+
+            if (test_random() % 5 == 0)
+                /* drop packet */
+                continue;
+
+            if (off <= queued_min && off + size > queued_min)
+                queued_min = off + size;
+
+            if (!TEST_true(ossl_quic_rstream_queue_data(rstream, NULL, off,
+                                                        bulk_data + off,
+                                                        size, 0)))
+                goto err;
+
+            if (test_random() % 5 != 0)
+               continue;
+
+            /* random overlapping retransmit */
+            off = read_off + test_random() % 50;
+            if (off > 50)
+                off -= 50;
```

Comment:
I don't understand the purpose of this check....can you explain it?

Reply:
This randomly moves the offset of the retransmitted data before the read_off so it simulates the situation where we receive data that was already read.

Reply:
Well - not really randomly at all. Except in the time through the loop when `r == 0` `off` will be `>50` because `read_off > 50`. When `r == 0` `off` is never `> 50`. Or am I reading something wrong here?

Reply:
The intention is for the final off to be somewhere between read_off - 50 and read_off - 1 except for the initial cycles where off would be negative which we cannot do.

Reply:
I meant randomly in regards to the whole computation.

Reply:
Ok. I understand now.